__stories__/FilteringVariableSizeTree.story.tsx

```diff
@@ -0,0 +1,202 @@
+/* eslint-disable max-depth */
+import {number, withKnobs} from '@storybook/addon-knobs';
+import {storiesOf} from '@storybook/react';
+import React, {FC, useState, useCallback, useEffect, useRef} from 'react';
+import AutoSizer from 'react-virtualized-auto-sizer';
+import {
+  TreeWalker,
+  TreeWalkerValue,
+  VariableSizeNodeData,
+  VariableSizeNodePublicState,
+  VariableSizeTree,
+} from '../src';
+import {NodeComponentProps} from '../src/Tree';
+
+document.body.style.margin = '0';
+document.body.style.display = 'flex';
+document.body.style.minHeight = '100vh';
+
+const root = document.getElementById('root')!;
+root.style.margin = '10px 0 0 10px';
+root.style.flex = '1';
+
+type TreeNode = Readonly<{
+  children: TreeNode[];
+  id: number;
+  name: string;
+}>;
+
+type NodeMeta = Readonly<{
+  nestingLevel: number;
+  node: TreeNode;
+}>;
+
+type ExtendedData = VariableSizeNodeData &
+  Readonly<{
+    isLeaf: boolean;
+    name: string;
+    nestingLevel: number;
+  }>;
+
+let nodeId = 0;
+
+const createNode = (depth: number = 0) => {
+  const node: TreeNode = {
+    children: [],
+    id: nodeId,
+    name: `test-${nodeId}`,
+  };
+
+  nodeId += 1;
+
+  if (depth === 5) {
+    return node;
+  }
+
+  for (let i = 0; i < 10; i++) {
+    node.children.push(createNode(depth + 1));
+  }
+
+  return node;
+};
+
+const rootNode = createNode();
+const defaultGapStyle = {marginLeft: 10};
+const defaultButtonStyle = {fontFamily: 'Courier New'};
+
+const Node: FC<NodeComponentProps<
+  ExtendedData,
+  VariableSizeNodePublicState<ExtendedData>
+>> = ({data: {isLeaf, name, nestingLevel}, isOpen, style, setOpen}) => {
+  return (
+    <div
+      style={{
+        ...style,
+        alignItems: 'center',
+        background: isLeaf ? '#fdd' : '#ddd',
+        display: 'flex',
+        marginLeft: nestingLevel * 30 + (isLeaf ? 48 : 0),
+      }}
+    >
+      {!isLeaf && (
+        <div>
+          <button
+            type="button"
+            onClick={() => setOpen(!isOpen)}
+            style={defaultButtonStyle}
+          >
+            {isOpen ? '-' : '+'}
+          </button>
+        </div>
+      )}
+      <div style={defaultGapStyle}>{name}</div>
+    </div>
+  );
+};
+
+type TreePresenterProps = Readonly<{
+  itemSize: number;
+}>;
+
+const getNodeData = (
+  node: TreeNode,
+  nestingLevel: number,
+): TreeWalkerValue<ExtendedData, NodeMeta> => ({
+  data: {
+    defaultHeight: node.children.length === 0 ? 30 : 60,
+    id: node.id.toString(),
+    isLeaf: node.children.length === 0,
+    isOpenByDefault: true,
+    name: node.name,
+    nestingLevel,
+  },
+  nestingLevel,
+  node,
+});
+
+let i = 0;
+function filterTree(tree: TreeNode, text: string) {
+  if (tree.children.length) {
```

Comment:
Thanks for this PR. 

What if there are multiple root nodes?

Reply:
my understanding is that when there are multiple root nodes, you would probably still have a top level root with tree.children.length, it is just that the tree walker doesn't yield the root level node https://github.com/Lodin/react-vtree/blob/master/__stories__/MultipleRoots.story.tsx#L84-L86

Reply:
Easy solution: 

Call `filterTree` for each root.

And instead of return your `template` when result is `null`:
`
{    name: 'no results',    children: [],    id: 'na',    nestingLevel: 0,  };
`

you can return `empty object` or `null` and filter it later from filtered result. It looks more better when you just hide empty nodes instead of comment each as 'no results'

Anyway thank you for this. I used it as base for my own filter solution.

Reply:
Any way to filter but keep the opened nodes? If we set isOpenByDefault to false (so we start with the tree not expanded), and we expand some nodes, and then filter, the nodes will not remain expanded. Any way to avoid this?

Reply:
Nervermind, we can use the async prop for the tree and it keeps the nodes open, don't mind me :)